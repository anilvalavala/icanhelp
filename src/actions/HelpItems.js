import database from '../../firebase/firebase';

export const getHelpItem = (id) => ({
    type: 'GET_HELP_ITEM',
    id
});

export const startGetHelpItem = (id) => {
    return (dispatch) => {
        return database.ref(`helpItems/${id}`)
                .once('value')
                .then((snapshot) => {
                    dispatch(getHelpItem(snapshot.key));
                });
    }
};

export const getAllHelpItems = (helpItems) => ({
    type: 'GET_ALL_ITEMS',
    helpItems
});

export const startGetAllHelpItems = () => {
    return (dispatch) => {
        return database.ref('helpItems')
                .once('value')
                .then((snapshot) => {
                    const helpItems = [];

                    snapshot.forEach((childSnapshot) => {
                        helpItems.push({
                            id: childSnapshot.key,
                            ...childSnapshot.val()
                        });
                    });
            dispatch(getAllHelpItems(helpItems))
        });
    };
};

export const addHelpItem = (helpItem) => 
({
    type: 'ADD_HELP_ITEM',
    helpItem
});

export const startAddHelpItem = (helpItemData = {}) => {
    return (dispatch) => {
        const {
            description = '',
            fromDate = 0,
            toDate = 0,
            title = '',
            email = '',
            phone = ''
        } = helpItemData;

        const helpItem = { description, fromDate, toDate, title, email, phone };

        //Here we're returning promise to allow promise chaining in test methods
        return database.ref('helpItems').push(helpItem).then((ref) => {
            dispatch(addHelpItem({
                id: ref.key,
                ...helpItem
            }));
        });
    };
};

export const editHelpItem = (helpItem) => ({
    type: 'EDIT_HELP_ITEM',
    helpItem
});

export const deleteHelpItem = (id) => ({
    type: 'DELETE_HELP_ITEM',
    id
});

export const startDeleteHelpItem = (id) => {
    return (dispatch) => {
        return database.ref(`helpItems/${id}`).remove().then(() => {
            dispatch(deleteHelpItem(id))
        });
    };
};
