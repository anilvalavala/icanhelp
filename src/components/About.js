import React from 'react';

const About = () => {
    return(
        <div>
            <h1>What is this about?</h1>
            <br />
            <p>
                Yes, today the world is full of sophisticated computers, machines, devices, food, what not, everything. <br />
                But still there are a section of people who are looking for help. <br /><br />
                There are many people who want to extend their helping hand to the society <br />
                But there is no bridge connecting such kind hearted people to the needy ones. <br /> <br />

                So, this application is intended for those people who want to donate something. <br />
                One can simply update their donation details and everyone on web can see it,  
                especially NGOs which are working for upliftment of less privileged. <br />

                This is just a basic verion of my idea. 
                Will keep updating this app to make it more user friendly.
            </p>
        </div>
    );
}

export default About;