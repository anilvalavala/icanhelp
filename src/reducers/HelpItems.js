import uuid from 'uuid';

const helpItemsReducer = (state = [{}], action) => {
    switch(action.type){
        case 'GET_HELP_ITEM':
            return state.filter(item => item.id === action.id);
        case 'ADD_HELP_ITEM':
            return [
                ...state,
                action.helpItem
            ];
        case 'EDIT_HELP_ITEM':
            return state.map((helpItem) => {
                if(helpItem.id == action.helpItem.id)
                {
                    return {
                        helpItem,
                        ...action.helpItem
                    }
                }else{
                    return helpItem;
                }
            });
        case 'DELETE_HELP_ITEM':
            return state.filter(item => item.id !== action.id);
        default:
            return state;
    }
}

export default helpItemsReducer;