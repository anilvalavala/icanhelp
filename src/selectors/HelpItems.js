import moment from 'moment';

const filteredHelpItems = (helpItems, filters) => {
    return helpItems.filter((helpItem) => {
            const toDateMoment = moment(helpItem.toDate)
            const filterByMatch =  !filters.filterBy ? true:  
                                   helpItem.description.toLowerCase()
                                   .includes(filters.filterBy.toLowerCase());

            const startDateMatch = filters.startDate ? 
                moment(filters.startDate).isSameOrBefore(toDateMoment, 'day') : 
                true
            
            const endDateMatch = filters.endDate ? 
                moment(filters.endDate).isSameOrAfter(toDateMoment, 'day') : 
                true

            return startDateMatch && endDateMatch && filterByMatch && helpItem.id;
        }).sort((a, b) => {
            if(filters.sortBy === 'fromDate'){
                return a.fromDate < b.fromDate ? 1 : -1
            }else if(filters.sortBy === 'toDate'){
                return a.toDate < b.toDate ? 1 : -1;
            }
        });
};

export default filteredHelpItems;