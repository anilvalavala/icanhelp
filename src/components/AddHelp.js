import React from 'react';
import { connect } from 'react-redux';
import HelpForm from './HelpForm';
import { startAddHelpItem } from '../actions/HelpItems';

export class AddHelp extends React.Component {
    onSubmit = (helpItem) => {
        this.props.startAddHelpItem(helpItem);
        this.props.history.push('/');
    };
    render(){
        return(
            <div>
                Add Help
                <HelpForm onSubmit={this.onSubmit} />
            </div>
        );
    };
}

// const AddHelp = (props) => {
//     return(
//         <div>
//         Add Help
//         <HelpForm onSubmit={(helpItem) => {
//             //The below line of code makes testing difficult
//             //Because addHelpItem is a dependency from another component
//             //props.dispatch(addHelpItem(helpItem));
//             props.onSubmit(helpItem)
//             props.history.push('/');
//         }}/>
//         </div>
//     );
// };

const mapDispatchToProps = (dispatch) => {
    return {
        startAddHelpItem: (helpItem) => dispatch(startAddHelpItem(helpItem))
    };
};


//When we do not have anything to connect from state to props but still we might need access to dispatch
//In this scenario we will use below syntax, connect without params
//export default connect()(AddHelp);

//connect has two params
//mapStateToProps - helps to attach state to props object, to consume in unconnected component
//mapDispatchToProps - helps to attach functions to props that can be used to dispatch actions
//mapDispatchToProps helps to abstract out dispatch calls from unconnected components
export default connect(undefined, mapDispatchToProps)(AddHelp);