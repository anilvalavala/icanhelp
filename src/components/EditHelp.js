import React from 'react';
import { connect } from 'react-redux';
import { startEditHelpItem } from '../actions/HelpItems';
import HelpForm from './HelpForm';

export class EditHelp extends React.Component {
    
    onSubmit = (helpItem) => {
        this.props.startEditHelpItem(helpItem);
        this.props.history.push('/home');
    };
    render(){
        return(
            <div>
            <HelpForm 
                helpItem={this.props.helpItem}
                onSubmit={this.onSubmit} />
        </div>
        );
    };
}

// const EditHelp = (props) => {
//     return(
//         <div>
//             <HelpForm 
//                 helpItem={props.helpItem}
//                 onSubmit={(helpItem) => {
//                     props.dispatch(editHelpItem(helpItem));
//                     props.history.push('/');
//             }} />
//         </div>
//     );
// };

const mapStateToProps = (state, props) => {
    return({
        helpItem: state.helpItems.find((helpItem) => {
            return helpItem.id === props.match.params.id
        })
    })
};

const mapDispatchToProps = (dispatch) => {
    return {
        startEditHelpItem: (helpItem) => dispatch(startEditHelpItem(helpItem))
    };
};

export default connect(mapStateToProps, mapDispatchToProps)(EditHelp);