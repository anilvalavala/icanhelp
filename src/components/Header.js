import React from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import { startLogout } from '../actions/Auth';

export const Header = ({ isAuthenticated, startLogout }) => {
    return(
    <div className="header">
    <div className="content-container">
        <Link className="header__link" to="/">Motive</Link>
        <Link className="header__link" to="/home">Find Help</Link>
        <Link className="header__link" to="/feedback">Feedback</Link>
        { !isAuthenticated && <Link className="header__link" to="/login">Login</Link> }
        <button className="header__link" onClick={startLogout}>Log Out</button>
    </div>
    </div>
    );
};

const mapStateToProps = (state) => ({
    isAuthenticated: !!state.auth.uid
});

const mapDispatchToProps = (dispatch) => ({
    startLogout: () => dispatch(startLogout())
});

export default connect(mapStateToProps, mapDispatchToProps)(Header);