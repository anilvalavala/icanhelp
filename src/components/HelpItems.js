import React from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import HelpItemDetails from './HelpItemDetails';
import filteredHelpItems from '../selectors/HelpItems';
import { deleteHelpItem } from '../actions/HelpItems';

//we are exporting unconnected version also to allow testing this component
export const HelpItems = (props) => {
    return(
        <div>
        {
            props.helpItems.length === 0 ? (
                <p>No Help Items Found</p>
            ) : (
                <div>
                <h1>Help Items</h1>
                {
                    props.helpItems.map(
                        (helpItem) => (
                            <div key={helpItem.id}>
                                <HelpItemDetails 
                                    helpItem={helpItem}
                                />
                                <br />
                            </div>
                        )
                    )
                }
                </div>
            )
        }
        </div>
    );
};

const mapStateToProps = (state) => ({
    helpItems: filteredHelpItems(state.helpItems, state.filters)
});

export default connect(mapStateToProps)(HelpItems);