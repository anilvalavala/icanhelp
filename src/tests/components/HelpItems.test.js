import React from 'react';
import { shallow } from 'enzyme';
import { HelpItems } from '../../components/HelpItems';
import TestHelpItems from '../fixtures/TestHelpItems';

test('Test HelpItems component by passing HelpItems', () => {
    const wrapper = shallow(<HelpItems helpItems={TestHelpItems} />);
    expect(wrapper).toMatchSnapshot();
});

test('Test HelpItems component without passing helpItems', () => {
    const wrapper = shallow(<HelpItems helpItems={[]} />);
    expect(wrapper).toMatchSnapshot();
});