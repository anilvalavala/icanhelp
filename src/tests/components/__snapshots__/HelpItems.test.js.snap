// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test HelpItems component by passing HelpItems 1`] = `
<div>
  <div>
    <h1>
      Help Items
    </h1>
    <div
      key="1"
    >
      <Connect(HelpItemDetails)
        helpItem={
          Object {
            "description": "Food available",
            "email": "anilkumar.v@hotmail.com",
            "fromDate": -86400000,
            "id": "1",
            "phone": "9545947627",
            "title": "Food",
            "toDate": 86400000,
          }
        }
      />
      <br />
    </div>
    <div
      key="2"
    >
      <Connect(HelpItemDetails)
        helpItem={
          Object {
            "description": "K.G books",
            "email": "anilkumar.v@hotmail.com",
            "fromDate": 86400000,
            "id": "2",
            "phone": "9545947627",
            "title": "books",
            "toDate": 259200000,
          }
        }
      />
      <br />
    </div>
    <div
      key="3"
    >
      <Connect(HelpItemDetails)
        helpItem={
          Object {
            "description": "Clothes",
            "email": "anilkumar.v@hotmail.com",
            "fromDate": 345600000,
            "id": "3",
            "phone": "9545947627",
            "title": "Clothes",
            "toDate": 518400000,
          }
        }
      />
      <br />
    </div>
  </div>
</div>
`;

exports[`Test HelpItems component without passing helpItems 1`] = `
<div>
  <p>
    No Help Items Found
  </p>
</div>
`;
